File: borrow/views/entity_list.py
from ._base import ApiView, Pagination, CheckPerm
from ..perm_types import entity_list_perm, revert_list_perm
from ..models.entity_list import EntityList
from ..models.revert_list import RevertList


class EntityListApiView(Pagination, ApiView):

    """有关实体清单的接口"""

    _search_fields = (
        ('status', 'status'),
        ('signer_id', 'signer_id'),
        ('sign_date', 'sign_date__range'),
        ('company_id', 'company_id'),
        ('descript', 'descript'),
        ('order_id', 'ordeer_id__contains'),
        ('entity_id', 'entity_id'),
        ('borrow_id', 'borrow_id'),
        ('borrow_date', 'borrow_date__range'),
        ('revert_date', 'revert_date__range')
    )
    _default_order = '-id'
    _list_fields = (
        'id',
        ('entity#name', 'entity'),
        'entity_id',
        'amount',
        'signer',
        'signer_id',
        'sign_date',
        'borrower',
        'borrower_id',
        'borrow_date',
        'revert_borrow_date',
        'revert_date',
        'status',
        'descript')

    @CheckPerm.check(entity_list_perm.view)
    def api_filter(self, request, args):
        filter_query = self.get_filter_query(args)
        search = EntityList.objects.filter(**filter_query)
        result = self.pagination(args, search)
        return self.success(result)

    @CheckPerm.check(entity_list_perm.view)
    def api_filter_all(self, request, args):
        filter_query = self.get_filter_query(args)
        status = filter_query.pop('status', None)
        search = EntityList.objects.filter(**filter_query)
        if status:
            search = search.exclude(status=status)
        result = self.decode(search)
        return self.success(result)

    @CheckPerm.check(entity_list_perm.add)
    def api_update(self, request, args):
        datas = args['datas']
        results = []

        for update in datas:
            model = {}
            for field in EntityList._meta.get_fields():
                name = field.name
                if field.is_relation:
                    name = name + '_id'
                if name not in update:
                    continue
                model[name] = update[name]

            if model.get('id'):
                result = model.pop('id')
                EntityList.objects.filter(pk=result).update(**model)
            else:
                entity_list = EntityList.objects.create(**model)
                entity_list.save()
                result = entity_list.id

            results.append(result)

        return self.success(results)

    @CheckPerm.check(entity_list_perm.add)
    def api_revert(self, request, args):
        if not args.get('selected'):
            return self.failed('没有选择归还物品', 2)

        # 生成订单编号
        order_id = RevertList.create(args['company_id'])

        EntityList.objects.filter(
            pk__in=args['selected']).update(
            status='归还', order_id=order_id)

        return self.success(order_id)

    @CheckPerm.check(revert_list_perm.view)
    def api_get_revert(self, request, args):
        revert_list_id = args.get('id')
        if not revert_list_id:
            return self.failed('参数错误', 3)

        search = EntityList.objects.filter(order_id_id=revert_list_id)
        result = self.decode(search)
        return self.success(result)

File: appfront/src/api/entity_list.js
import BaseApi from './base.js'
import {createFakePageData, createFakeData} from '../tools.js'

const orderId = 'YH201804010010'

let datas = [
  {id: 1, entity: '证明', amount: 1, signer: 'admin', sign_date: '2018-03-08', borrower: '雇员1', borrow_date: '2018-03-11', revert_borrow_date: '2018-04-11', revert_date: '2018-04-23', status: '寄存', descript: 'test', entity_id: 3, signer_id: 1, borrower_id: 2, company_id: 1},
  {id: 4, entity: '身份证', amount: 1, signer: 'admin', sign_date: '2018-03-08', borrower: '雇员1', borrow_date: '2018-03-11', revert_borrow_date: '2018-04-11', revert_date: '2018-04-23', status: '寄存', descript: 'test', entity_id: 3, signer_id: 1, borrower_id: 2, company_id: 1},
  {id: 5, entity: '证明', amount: 1, signer: 'admin', sign_date: '2018-03-08', borrower: '雇员1', borrow_date: '2018-03-11', revert_borrow_date: '2018-04-11', revert_date: '2018-04-23', status: '寄存', descript: 'test', entity_id: 3, signer_id: 1, borrower_id: 2, company_id: 1},
  {id: 6, entity: '证明', amount: 1, signer: 'admin', sign_date: '2018-03-08', borrower: '雇员1', borrow_date: '2018-03-11', revert_borrow_date: '2018-04-11', revert_date: '2018-04-23', status: '寄存', descript: 'test', entity_id: 3, signer_id: 1, borrower_id: 2, company_id: 1},
  {id: 7, entity: '身份证', amount: 1, signer: 'admin', sign_date: '2018-03-08', borrower: '雇员1', borrow_date: '2018-03-11', revert_borrow_date: '2018-04-11', revert_date: '2018-04-23', status: '寄存', descript: 'test', entity_id: 3, signer_id: 1, borrower_id: 2, company_id: 1, order_id: orderId},
  {id: 2, entity: '护照', amount: 2, signer: '雇员2', sign_date: '2018-03-08', borrower: '雇员1', borrow_date: '2018-03-11', revert_borrow_date: '2018-04-11', revert_date: '2018-04-23', status: '归还', descript: 'test', entity_id: 2, signer_id: 3, borrower_id: 2, company_id: 1, order_id: orderId},
  {id: 3, entity: '证明', amount: 2, signer: '雇员2', sign_date: '2018-03-08', borrower: '雇员1', borrow_date: '2018-03-11', revert_borrow_date: '2018-04-11', revert_date: '2018-04-23', status: '归还', descript: 'test', entity_id: 2, signer_id: 3, borrower_id: 2, company_id: 2, order_id: orderId}
]

class EntityListApi extends BaseApi {
  fakeApiData (args) {
    switch (args.type) {
      case 'api_get_revert':
        return createFakeData(datas.filter(x => x.order_id === args.order_id))
      case 'api_filter':
        return createFakePageData(datas.filter(x => x.company_id === args.company_id && x.status === args.status))
      case 'api_filter_all':
        return createFakeData(datas.filter(x => x.company_id === args.company_id && x.status !== args.status))
      case 'api_update':
        datas = Array.prototype.concat.call([], datas, args.datas)
        return createFakeData(true)
      case 'api_revert':
        const found = datas.filter(x => args.selected.includes(x.id))
        if (found) {
          found.forEach(x => {
            x['status'] = '归还'
            x['order_id'] = orderId
          })
          return createFakeData(orderId)
        } else {
          return this.fakeApiData({type: 'default'})
        }
      default:
        return {status: false, code: -100, msg: ''}
    }
  }

  filter (args) {
    return this.fetch({type: 'api_filter', ...args})
  }

  filterAll (args) {
    return this.fetch({type: 'api_filter_all', ...args})
  }

  update (args) {
    return this.fetch({type: 'api_update', ...args})
  }

  revert (args) {
    return this.fetch({type: 'api_revert', ...args})
  }

  getRevert (args) {
    return this.fetch({type: 'api_get_revert', ...args})
  }
}

const entityListApi = new EntityListApi('/borrow/api/entity_list')

export default entityListApi
